arch-riscv: Fix misprediction of control flow instruction caused by vset{i}vl{i}

5265325 opened the pull request:
This PR fixed the bug mentioned in #1708.

In the new implementation, the BPU only predicts vl&vtype for vset{i}vl{i} instructions. Other control flow instructions will not be predicted by the BPU for vl and vtype, because it is meaningless. The new implementation adds `new_conf` flag in pcstate and `vl/vtype` registers in decoder. This is used to filter the prediction of `vl&vtype` by other control flow instructions. When a flush occurs, the `vl/vtype` registers in the decoder are restored.

When checking pcstate at IEW stage, `vl&vtype` are only checked for instructions with new_vconf = 1. It means that vl and vtype are only checked for vset{i}vl{i} instructions.

Here is a simple picture describing the new changes：
![绘图1](https://github.com/user-attachments/assets/210e1d34-00d2-4111-a389-e1306d8f25e1)

The following picture is the new pipeline of the test case from #1708. No more flush after warmup.
![截屏2024-10-24 下午4 08 02](https://github.com/user-attachments/assets/9942b2ef-73f2-463b-a982-74950f0e726a)

Change-Id: I7d3df0328d5757f6b3506743200d5b154dee367f

Commit:
arch-riscv: Fix misprediction of control flow instruction caused by vset{i}vl{i}

Change-Id: I7d3df0328d5757f6b3506743200d5b154dee367f

saul44203 reviewed (commented):
I think we can remove `BranchTarget` for both `vsetvli` and `vsetvl` since they are `IsIndirectControl` and will not execute this anyways (only `vsetivli` really needs it). Apart from this it looks good to me.

erin-le commented:
@5265325 Hello, thanks for your contribution! Could you respond to the comment on this change?

saul44203 commented:
Hi @5265325. Assuming this is done and tested, LGTM.

hnpl commented on src/arch/riscv/pcstate.hh:
```diff
@@ -120,15 +126,16 @@ class PCState : public GenericISA::UPCState<4>
     {
         Base::serialize(cp);
         SERIALIZE_SCALAR(_rvType);
+        SERIALIZE_SCALAR(_new_vconf);
```
@Harshil2107 Does this require updating the checkpointing upgrade script?

5265325 replied:
Currently the iscv-hello-example-checkpoint test uses checkpoint from http://dist.gem5.org/dist/develop/checkpoints/riscv-hello-example-checkpoint.tar. But the checkpoint does not contain `_new_vconf` entry. So it cannot pass the test in CI.

hnpl reviewed (approved):
The code looks good to me.

We probably need update the checkpoint upgrader but it can be done in a different commit.

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -4845,6 +4846,7 @@ decode QUADRANT default Unknown::unknown() {
                             Rd_ud = new_vl;
                             VL = new_vl;
                             Vtype = new_vtype;
+                            IsVset = true;
                         }}, VSetVlDeclare, VSetVlBranchTarget
                           , SimdConfigOp, IsUncondControl
                           , IsIndirectControl);
```
I noticed you removed the `branchTarget` for `vsetvl` and `vsetvli`. Not sure if we need to remove the instruction flag `IsIndirectControl` and `IsUncondControl` and mark them `IsSerializeAfter` and `IsNonSpeculative`. Not sure if there will be the issue if BPU support indirect jump prediction in the future.

5265325 replied:
I had similar thoughts. At least vsetivli instructions don't need to be speculated on. I think we can create a new pr after the current pr is merged to modify the code and discuss it.

powerjg commented:
@Harshil2107 can you work with @5265325 to update the checkpoints for the CI tests?

(Note: checkpoint updates should probably be included in the release notes)

Harshil2107 commented:
I tested the checkpoint upgrader, and it works fine. However, the updated section needs to be in a new file because modifying an existing upgrader file won’t reapply changes to checkpoints that already have the tag from the previous version. Is it okay if I update the upgrader and push a commit?

5265325 commented:
> I tested the checkpoint upgrader, and it works fine. However, the updated section needs to be in a new file because modifying an existing upgrader file won’t reapply changes to checkpoints that already have the tag from the previous version. Is it okay if I update the upgrader and push a commit?

Thank you for testing the checkpoint upgrader and providing feedback! Feel free to do it if you need to.

Harshil2107 committed:
util: Move cpt upgrader to new file

- Moved the change made to the `riscv-pcstate.py` upgrader to a new file, `riscv-vconf.py`.
- Updated `riscv-hello-restore-checkpoint.py` to reference the new checkpoint file.
